### PR TITLE
Best support vector tiles

### PR DESCRIPTION
**Pull request by gchoqueux**
## Description
Add supports to
 * layers ordering
 * pattern
 * fill stroke
 * line cap and join
 * dash line
 * multi styles
 * styles by zoom level

and some fixes.

vt

### CODE REVIEW

File: src/Source/Source.js [hunk 1]
```diff
@@ -116,6 +116,10 @@ class Source {
         }
     }
 
+    isVectorSource() {
+        return supportedParsers.get(this.format) !== undefined;
```
Comment (zarov) [inline src/Source/Source.js]: This could be a value set in the constructor, as it doesn't seem to change.

File: src/Core/Feature.js [hunk 2]
```diff
@@ -291,4 +291,26 @@ export class FeatureCollection {
             return newFeature;
         }
     }
+
+    getFeatureById(id, type) {
+        const feature = this.features.find(feature => feature.id === id);
+        if (feature && this.optionsFeature.mergeFeatures) {
```
Comment (zarov) [inline src/Core/Feature.js]: If I understand correctly, simply having `feature` is not sufficient and will return a new one ? Why ?

File: src/Core/Feature.js [hunk 3]
```diff
@@ -291,4 +291,26 @@ export class FeatureCollection {
...
+            return feature;
+        } else {
+            const newFeature = new Feature(type, this.crs, this.optionsFeature);
+            this.features.push(newFeature);
+            return newFeature;
+        }
+    }
+
+    getReference(feature) {
```
Comment (zarov) [inline src/Core/Feature.js]: Could you add a little comment/documentation on what is this method supposed to return ? It seems that it is a `clone`, so maybe do a `Feature#clone` method instead ?

File: src/Core/Style.js [hunk 4]
```diff
@@ -88,35 +101,58 @@ class Style {
...
+            const { color, opacity } = rgba2rgb(readVectorProperty(layer.paint['fill-color'] || layer.paint['fill-pattern']));
             this.fill.color = color;
-            this.fill.opacity = layer.paint['fill-opacity'] || opacity || 1.0;
+            this.fill.opacity = readVectorProperty(layer.paint['fill-opacity'], zoom) || opacity || 1.0;
+            if (layer.paint['fill-pattern']) {
+                const sprite = sprites[layer.paint['fill-pattern']];
+                const c = document.createElement('canvas');
+                c.width = sprite.width;
+                c.height = sprite.height;
+                const ctx = c.getContext('2d');
```
Comment (zarov) [inline src/Core/Style.js]: Could you instance a canvas globally, so you don't have to get the context everytime ?

File: src/Parser/VectorTileParser.js [hunk 5]
```diff
@@ -138,25 +139,44 @@ function readPBF(file, options) {
...
-            const layerStyle = layersStyle.filter(l => l.filterExpression({ zoom: extentSource.zoom }, vtFeature))[0];
-            if (layerStyle) {
-                const properties = vtFeature.properties;
-                properties.style = styleCache.get(layerStyle.id);
-                if (!properties.style) {
-                    properties.style = new Style();
-                    properties.style.setFromVectorTileLayer(layerStyle);
-                    styleCache.set(layerStyle.id, properties.style);
+            const layerStyles = layersStyle.filter(l => l.filterExpression({ zoom: z }, vtFeature));
+            const layerStyle = layerStyles.filter(l => (!l.minzoom || l.minzoom <= z) && (!l.maxzoom || l.maxzoom >= z));
```
Comment (zarov) [inline src/Parser/VectorTileParser.js]: Why not apply all conditions to the filter the first time ?

**Review by zarov — CHANGES_REQUESTED** [review]
Nice to have those improvements !

File: src/Core/Feature.js (hunk 2 above)
Comment (gchoqueux) [reply src/Core/Feature.js]: yes

File: src/Source/Source.js (hunk 1 above)
Comment (gchoqueux) [reply src/Source/Source.js]: yes

File: src/Core/Feature.js (hunk 2 above)
Comment (gchoqueux) [reply src/Core/Feature.js]: For rendering, Vector tiles layers need to be ordered. If layers are grouped by type, it's more complicated to order. So, for vector tiles I add a feature/reference by layer.

File: src/Core/Feature.js (hunk 3 above)
Comment (gchoqueux) [reply src/Core/Feature.js]: I added a documentation

File: src/Core/Style.js (hunk 4 above)
Comment (gchoqueux) [reply src/Core/Style.js]: I'll try

File: src/Parser/VectorTileParser.js (hunk 5 above)
Comment (gchoqueux) [reply src/Parser/VectorTileParser.js]: yes

**Comment by gchoqueux** [comment]
@zarov PR updated

File: src/Source/Source.js [hunk 6]
```diff
@@ -106,6 +106,7 @@ class Source {
         this.format = source.format;
         this.fetcher = source.fetcher || supportedFetchers.get(source.format) || Fetcher.texture;
         this.parser = source.parser || supportedParsers.get(source.format) || (d => Promise.resolve(d));
+        this.isVectorSource = source.parser || supportedParsers.get(source.format);
```
Comment (zarov) [inline src/Source/Source.js]: This isn't a boolean. It might works for some case, but it's better to set a boolean.

File: src/Core/Feature.js [hunk 7]
```diff
@@ -273,6 +273,16 @@ export class FeatureCollection {
         this.features.push(feature);
         this.updateExtent(feature.extent);
     }
+    getFeature(type, callback) {
+        const feature = this.features.find(callback);
+        if (feature && this.optionsFeature.mergeFeatures) {
```
Comment (zarov) [inline src/Core/Feature.js:279]: It should return the `feature` if it is found, if `mergeFeatures` is not defined, so you get something like this:
```js
if (feature && (this.optionsFeature.mergeFeatures == undefined || this.optionsFeature.mergeFeatures)) {
```
or else you should rename the function as it does not really get the Feature but only the Feature if the Features are merged.

File: examples/js/plugins/FeatureToolTip.js [hunk 8]
```diff
@@ -75,7 +75,7 @@ var FeatureToolTip = (function _() {
             feature = features[p];
 
             geometry = feature.geometry;
-            style = (layer.style && layer.style.isStyle) ? layer.style : geometry.properties.style;
+            style = (layer.style && layer.style.isStyle) ? layer.style : (feature.geometry._feature.style || geometry.properties.style);
```
Comment (zarov) [inline examples/js/plugins/FeatureToolTip.js]: Can be `geometry._feature.style`

**Review by zarov — APPROVED** [review]
LGTM, thanks for these improvements !

**Pull request merged**